ftw/http.py

```diff
@@ -463,38 +462,77 @@ def get_response(self):
+                            'proto': self.request_object.protocol,
+                            'msg': 'Please send the request and check'
+                            + ' Wireshark',
+                            'function': 'http.HttpUA.get_response'
+                        })
+
+    def read_response_from_socket(self):
+        # wait for socket to become ready
+        ready_sock, _, _ = select.select(
+            [self.sock], [], [self.sock], self.SOCKET_TIMEOUT)
+        if not ready_sock:
+            raise errors.TestError(
+                f'No response from server within {self.SOCKET_TIMEOUT}s',
+                {
+                    'host': self.request_object.dest_addr,
+                    'port': self.request_object.port,
+                    'proto': self.request_object.protocol,
+                    'msg': 'Please send the request and check Wireshark',
+                    'function': 'http.HttpUA.get_response'
+                })
+
+        our_data = []
         while True:
-            # If we have data then if we're passed the timeout break
-            if our_data and time.time() - begin > self.HTTP_TIMEOUT:
-                break
-            # If we're dataless wait just a bit
-            elif time.time() - begin > self.HTTP_TIMEOUT * 2:
-                break
-            # Recv data
             try:
                 data = self.sock.recv(self.RECEIVE_BYTES)
-                if data:
-                    our_data.append(util.ensure_binary(data))
-                    begin = time.time()
-                else:
-                    # Sleep for sometime to indicate a gap
-                    time.sleep(self.HTTP_TIMEOUT)
-            except socket.error as err:
-                # Check if we got a timeout
-                if err.errno == errno.EAGAIN:
-                    pass
+                if len(data) == 0:
+                    # we're done
+                    break
+                our_data.append(util.ensure_binary(data))
+            except BlockingIOError as e:
+                # If we can't handle the error here, pass it on
+                if e.errno == socket.EAGAIN or e.errno == socket.EWOULDBLOCK:
+                    # we're done
+                    break
+            except OSError as err:
                 # SSL will return SSLWantRead instead of EAGAIN
-                elif sys.platform == 'win32' and \
-                        err.errno == errno.WSAEWOULDBLOCK:
+                if (sys.platform == 'win32' and
+                        err.errno == errno.WSAEWOULDBLOCK):
                     pass
                 elif (self.request_object.protocol == 'https' and
-                      err.args[0] == ssl.SSL_ERROR_WANT_READ):
-                    continue
-                # If we didn't it's an error
+                        err.args[0] == ssl.SSL_ERROR_WANT_READ):
+                    ready_sock, _, _ = select.select(
+                        [self.sock], [], [self.sock], .3)
```

Comment:
In the end, this `.3` is the self.HTTP_TIMEOUT from above, right? Why not make it explicit? Looks really hidden here.

Reply:
Yes, you're right.